### ISSUE TITLE
window.reportError(e)?

### ISSUE THREAD

**Issue by benjamingr** [denoland/deno#13484]
The web platform supports `self.reportError(e)` to trigger an uncaught error with semantics similar to `setTimeout(() => { throw e; })`.

Would it make sense for Deno to implement this API?

**Comment by bartlomieju** [comment]
Looking at [mdn](https://developer.mozilla.org/en-US/docs/Web/API/reportError) it appears that this API is only available in Worker context.

**Comment by andreubotella** [comment]
It isn't. It's _also_ available in workers.

**Comment by benjamingr** [comment]
Yes if you look at the [spec](https://html.spec.whatwg.org/multipage/webappapis.html#windoworworkerglobalscope-mixin) it's exposed in both cases.

(I am not sure if adding it is the thing to do - we decided not to implement this API (yet) in Node.js and the suggestion to add it stalled)

**Comment by bartlomieju** [comment]
Okay, this seems useful. I'm in favor of adding this API. If anyone is interested in working on a PR, let me know, we'll try to provide some pointers.

**Comment by benjamingr** [comment]
> @benjamingr Do I have this correct with respect to what the Node behaviour would be + process termination?

You do, though I still haven't merged the Node.js PR since I am still not 100% sure this is a good API.

Also "The error is printed to the console in the format of an exception." - this likely won't happen if there is an errorMonitor or uncaughtException handler so the user can redirect these to monitoring.

**Comment by nayeemrmn** [comment]
> Also "The error is printed to the console in the format of an exception." - this likely won't happen if there is an errorMonitor or uncaughtException handler so the user can redirect these to monitoring.

Perfect, yeah I meant that to be conditioned under the unhandled case as well since browsers already do that with `e.preventDefault()`.

Looking at https://github.com/nodejs/node/pull/41912 it looks like we reached the same conclusion to crash by throwing on the next tick. One difference however is that by elevating the error for the first time at the next tick means it won't be printed synchronously like it is in browsers.

<details>
<summary>I avoided this by splitting the error into two, the input error is printed immediately while a generic error is thrown next tick to cause the crash:</summary>

```js
console.log(1);
reportError(new Error("foo"));
console.log(2);
```
![image](https://user-images.githubusercontent.com/29990554/156128385-61415db7-9236-4ef9-8150-d8fa264953a3.png)
</details>
<details>
<summary>Probably wouldn't be as palatable if we didn't already have this kind of formatting for child worker errors:</summary>

```ts
// temp.ts
new Worker(new URL("temp2.ts", import.meta.url).href, { type: "module" });

// temp2.ts
throw new Error("foo");
```
![image](https://user-images.githubusercontent.com/29990554/156128694-6b5bdb5c-5a8c-4c74-88a1-bf9af59a425d.png)
</details>

Ref https://github.com/denoland/deno/blob/3dec3301108a3dc85ffd99c5e60eb5ab613150b0/ext/web/16_report_error.js

**Comment by benjamingr** [comment]
What do you think should happen in terms of developer experience when someone does `reportError("error")` (that is - something without a stack trace)?

We are discussing either warning in Node.js (a non-error causes a warning with a stack trace), always logging the stack trace when `.reportError` is called at the call site or doing nothing.

**Comment by nayeemrmn** [comment]
@benjamingr Maybe Node should tie it to `--trace-uncaught` and essentially just match the output of `throw "error"`.

```
$ node -e 'reportError("error")'
[eval]:1
reportError("error")
^
1
(Use `node --trace-uncaught ...` to show where `reportError()` was called)

Node.js vX.X.X

$ node --trace-uncaught -e 'reportError("error")'
[eval]:1
reportError("error")
^
1
`reportError()` called at:
    at [eval]:1:1
    at runInThisContext (node:vm:129:12)
    at runInThisContext (node:vm:305:38)
    at node:internal/process/execution:76:19
    at [eval]-wrapper:6:22
    at evalScript (node:internal/process/execution:75:60)
    at node:internal/main/eval_string:27:3

Node.js vX.X.X
```

**Comment by lucacasonato** [comment]
Will be released in Deno 1.21.0
